pkg/controller/multi_unmapper_at.go

```diff
@@ -0,0 +1,78 @@
+package controller
+
+import (
+	"strings"
+	"sync"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/longhorn/longhorn-engine/pkg/types"
+)
+
+type MultiUnmapperAt struct {
+	lock      *sync.Mutex
+	unmappers []types.UnmapperAt
+}
+
+type MultiUnmapperError struct {
+	errors []error
+}
+
+func (m *MultiUnmapperError) Error() string {
+	errors := []string{}
+	for _, err := range m.errors {
+		if err != nil {
+			errors = append(errors, err.Error())
+		}
+	}
+
+	switch len(errors) {
+	case 0:
+		return "Unknown"
+	case 1:
+		return errors[0]
+	default:
+		return strings.Join(errors, "; ")
+	}
+}
+
+func (m *MultiUnmapperAt) UnmapAt(length uint32, off int64) (int, error) {
+	errs := make([]error, len(m.unmappers))
+	wg := sync.WaitGroup{}
+
+	unmappedSize := 0
+
+	for i, u := range m.unmappers {
+		wg.Add(1)
+		go func(index int, u types.UnmapperAt) {
+			n, err := u.UnmapAt(length, off)
+			if err != nil {
+				errs[index] = err
+			}
+
+			m.lock.Lock()
+			if unmappedSize == 0 {
+				unmappedSize = n
+			} else if unmappedSize != n {
+				logrus.Warnf("One of the MultiUnmapper get different size %v from others %v", n, unmappedSize)
+			}
+			m.lock.Unlock()
+
+			wg.Done()
+		}(i, u)
+	}
+
+	wg.Wait()
+
+	var err error = nil
+	for i := range errs {
+		if errs[i] != nil {
```

Comment:
You already make sure `errs` will not have nil value above, so can just create and return a `MultiUnmapperError` if `len(errs) > 0` instead of using this needless loop.

Reply:
I cannot apply this change: `len(errs)` is always non-zero value: `errs := make([]error, len(m.unmappers))`. 
As a result, some of the entries are indeed nil.

Reply:
oh, I see, then need to filter.